File: packages/sdk-go/budget_attestation_test.go
package rateguard

import (
	"crypto/ed25519"
	"crypto/rand"
	"testing"
	"time"
)

func genKey(t *testing.T) (ed25519.PublicKey, ed25519.PrivateKey) {
	t.Helper()
	pub, priv, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		t.Fatal(err)
	}
	return pub, priv
}

func rootGrant(now time.Time) BudgetGrant {
	return BudgetGrant{
		MaxTokens: 1_000_000,
		Providers: []string{"openai", "anthropic"},
		Models:    []string{"gpt-4o", "claude-opus-4-5"},
		MaxDepth:  3,
		ExpiresAt: now.Add(time.Hour),
	}
}

func TestNewRootBudgetTokenAndVerifyChain(t *testing.T) {
	authorityPub, authorityPriv := genKey(t)
	now := time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)

	token, delegatePriv, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: rootGrant(now)})
	if err != nil {
		t.Fatal(err)
	}
	if delegatePriv == nil {
		t.Fatal("expected a generated delegate private key")
	}
	if len(token.Blocks) != 1 {
		t.Fatalf("expected 1 block, got %d", len(token.Blocks))
	}

	grant, err := verifyChainAt(token, authorityPub, now)
	if err != nil {
		t.Fatalf("verify chain: %v", err)
	}
	if grant.MaxTokens != 1_000_000 {
		t.Fatalf("effective grant mismatch: %+v", grant)
	}
}

func TestVerifyChainRejectsWrongRootKey(t *testing.T) {
	_, authorityPriv := genKey(t)
	wrongPub, _ := genKey(t)
	now := time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)

	token, _, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: rootGrant(now)})
	if err != nil {
		t.Fatal(err)
	}
	if _, err := verifyChainAt(token, wrongPub, now); err == nil {
		t.Fatal("expected verification to fail against the wrong root public key")
	}
}

func TestAttestSingleHopDelegation(t *testing.T) {
	authorityPub, authorityPriv := genKey(t)
	now := time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)

	root, rootPriv, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: rootGrant(now)})
	if err != nil {
		t.Fatal(err)
	}

	delegated, delegatePriv, err := Attest(root, rootPriv, AttestOptions{Grant: BudgetGrant{
		MaxTokens: 10_000, // narrower
		Providers: []string{"openai"},
		Models:    []string{"gpt-4o"},
		MaxDepth:  1, // narrower (<= 3-1)
		ExpiresAt: now.Add(30 * time.Minute),
	}})
	if err != nil {
		t.Fatalf("attest: %v", err)
	}
	if delegatePriv == nil {
		t.Fatal("expected a generated delegate private key for the sub-agent")
	}
	if len(delegated.Blocks) != 2 {
		t.Fatalf("expected a 2-block chain, got %d", len(delegated.Blocks))
	}

	grant, err := verifyChainAt(delegated, authorityPub, now)
	if err != nil {
		t.Fatalf("verify delegated chain: %v", err)
	}
	if grant.MaxTokens != 10_000 {
		t.Fatalf("effective grant should be the narrowed leaf, got %+v", grant)
	}
}

func TestAttestRejectsWideningTokens(t *testing.T) {
	_, authorityPriv := genKey(t)
	now := time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)
	root, rootPriv, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: rootGrant(now)})
	if err != nil {
		t.Fatal(err)
	}

	_, _, err = Attest(root, rootPriv, AttestOptions{Grant: BudgetGrant{
		MaxTokens: 2_000_000, // wider than parent's 1,000,000
		MaxDepth:  1,
		ExpiresAt: now.Add(30 * time.Minute),
	}})
	if err == nil {
		t.Fatal("expected rejection of a wider MaxTokens grant")
	}
}

func TestAttestRejectsWideningProviders(t *testing.T) {
	_, authorityPriv := genKey(t)
	now := time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)
	root, rootPriv, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: rootGrant(now)})
	if err != nil {
		t.Fatal(err)
	}

	_, _, err = Attest(root, rootPriv, AttestOptions{Grant: BudgetGrant{
		MaxTokens: 100,
		Providers: []string{"openai", "google"}, // "google" not in parent's [openai, anthropic]
		MaxDepth:  1,
		ExpiresAt: now.Add(30 * time.Minute),
	}})
	if err == nil {
		t.Fatal("expected rejection of a provider list widening beyond the parent's set")
	}
}

func TestAttestRejectsLaterExpiry(t *testing.T) {
	_, authorityPriv := genKey(t)
	now := time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)
	root, rootPriv, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: rootGrant(now)})
	if err != nil {
		t.Fatal(err)
	}

	_, _, err = Attest(root, rootPriv, AttestOptions{Grant: BudgetGrant{
		MaxTokens: 100,
		MaxDepth:  1,
		ExpiresAt: now.Add(2 * time.Hour), // later than parent's 1 hour
	}})
	if err == nil {
		t.Fatal("expected rejection of an expiry later than the parent's")
	}
}

func TestAttestRejectsDepthExhaustion(t *testing.T) {
	_, authorityPriv := genKey(t)
	now := time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)
	root, rootPriv, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: BudgetGrant{
		MaxTokens: 1000,
		MaxDepth:  0, // no further delegation allowed
		ExpiresAt: now.Add(time.Hour),
	}})
	if err != nil {
		t.Fatal(err)
	}

	_, _, err = Attest(root, rootPriv, AttestOptions{Grant: BudgetGrant{
		MaxTokens: 100,
		MaxDepth:  0,
		ExpiresAt: now.Add(30 * time.Minute),
	}})
	if err == nil {
		t.Fatal("expected rejection of delegation from a depth-exhausted token")
	}
}

func TestAttestRejectsWrongParentKey(t *testing.T) {
	_, authorityPriv := genKey(t)
	_, wrongPriv := genKey(t)
	now := time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)
	root, _, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: rootGrant(now)})
	if err != nil {
		t.Fatal(err)
	}

	_, _, err = Attest(root, wrongPriv, AttestOptions{Grant: BudgetGrant{
		MaxTokens: 100,
		MaxDepth:  1,
		ExpiresAt: now.Add(30 * time.Minute),
	}})
	if err == nil {
		t.Fatal("expected rejection when the private key doesn't match the token's current holder")
	}
}

func TestVerifyChainRejectsExpiredToken(t *testing.T) {
	authorityPub, authorityPriv := genKey(t)
	now := time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)
	token, _, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: rootGrant(now)})
	if err != nil {
		t.Fatal(err)
	}

	future := now.Add(2 * time.Hour)
	if _, err := verifyChainAt(token, authorityPub, future); err == nil {
		t.Fatal("expected verification to fail once the grant has expired")
	}
}

// TestVerifyChainEnforcesTheSameTruncatedExpiryItSigned reproduces a real
// gap: the signature commits to ExpiresAt truncated to whole seconds (see
// signingPayload), but the expiry CHECK used to compare against the raw,
// untruncated value. A grant expiring at T.999 was signed as if it expired
// at T, but verifyChainAt would still accept it up to 999ms past T — the
// enforced statement was looser than the signed one.
func TestVerifyChainEnforcesTheSameTruncatedExpiryItSigned(t *testing.T) {
	authorityPub, authorityPriv := genKey(t)
	// .999999999s past the whole second — signingPayload truncates this to
	// exactly `base`, so the signature commits to `base`, not `base+999ms`.
	base := time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)
	grant := rootGrant(base)
	grant.ExpiresAt = base.Add(999999999 * time.Nanosecond)
	token, _, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: grant})
	if err != nil {
		t.Fatal(err)
	}

	// 500ms past `base`: still before the RAW expiry (base+999999999ns), so
	// the old buggy check (comparing against the raw value) would have
	// accepted this — but it's already after the TRUNCATED expiry the
	// signature actually committed to, so it must be rejected.
	if _, err := verifyChainAt(token, authorityPub, base.Add(500*time.Millisecond)); err == nil {
		t.Fatal("expected verification to reject a token past the truncated expiry instant the signature actually committed to, even though it's before the raw sub-second expiry")
	}

	// At exactly `base` (the truncated/signed instant itself), it must
	// still verify — not yet expired.
	if _, err := verifyChainAt(token, authorityPub, base); err != nil {
		t.Fatalf("expected verification to succeed exactly at the truncated expiry instant: %v", err)
	}
}

func TestSignAndVerifyPresentation(t *testing.T) {
	authorityPub, authorityPriv := genKey(t)
	// VerifyPresentation checks expiry against real wall-clock time (unlike
	// verifyChainAt, which the other tests in this file call directly with an
	// injected `now`), so the grant must be anchored to time.Now(), not a
	// fixed historical date that will eventually be in the past.
	now := time.Now()
	token, delegatePriv, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: rootGrant(now)})
	if err != nil {
		t.Fatal(err)
	}

	context := []byte("request-nonce-abc123")
	sig, err := Sign(token, delegatePriv, context)
	if err != nil {
		t.Fatalf("sign: %v", err)
	}

	grant, err := VerifyPresentation(token, authorityPub, context, sig)
	if err != nil {
		t.Fatalf("verify presentation: %v", err)
	}
	if grant.MaxTokens != 1_000_000 {
		t.Fatalf("unexpected grant: %+v", grant)
	}
}

func TestVerifyPresentationRejectsWrongHolder(t *testing.T) {
	_, authorityPriv := genKey(t)
	_, impostorPriv := genKey(t)
	now := time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)
	token, _, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: rootGrant(now)})
	if err != nil {
		t.Fatal(err)
	}

	context := []byte("request-nonce")
	// The impostor doesn't hold the token's delegate key, so Sign itself
	// must refuse — this is the defensive check inside Sign.
	if _, err := Sign(token, impostorPriv, context); err == nil {
		t.Fatal("Sign must reject a private key that doesn't match the token's holder key")
	}
}

func TestVerifyPresentationRejectsReplayedSignatureUnderDifferentContext(t *testing.T) {
	authorityPub, authorityPriv := genKey(t)
	// Same real-clock consideration as TestSignAndVerifyPresentation above:
	// VerifyPresentation checks expiry against time.Now(), so anchor to it.
	now := time.Now()
	token, delegatePriv, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: rootGrant(now)})
	if err != nil {
		t.Fatal(err)
	}

	sig, err := Sign(token, delegatePriv, []byte("original-context"))
	if err != nil {
		t.Fatal(err)
	}
	if _, err := VerifyPresentation(token, authorityPub, []byte("different-context"), sig); err == nil {
		t.Fatal("a signature over one context must not verify against a different context")
	}
}

func TestSignRejectsTokenNotMatchingDelegateKey(t *testing.T) {
	_, authorityPriv := genKey(t)
	now := time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)
	explicitDelegatePub, explicitDelegatePriv := genKey(t)

	token, generatedPriv, err := NewRootBudgetToken(authorityPriv, AttestOptions{
		Grant:             rootGrant(now),
		DelegatePublicKey: explicitDelegatePub,
	})
	if err != nil {
		t.Fatal(err)
	}
	if generatedPriv != nil {
		t.Fatal("no private key should be generated when DelegatePublicKey was supplied")
	}

	// The explicit delegate's own private key must work.
	if _, err := Sign(token, explicitDelegatePriv, []byte("ctx")); err != nil {
		t.Fatalf("explicit delegate should be able to sign: %v", err)
	}
}

func TestBudgetTokenMarshalRoundTripPreservesVerification(t *testing.T) {
	authorityPub, authorityPriv := genKey(t)
	now := time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)

	// Unrestricted root (nil Providers/Models) so the child can validly
	// leave them nil too — this exercises nil-vs-empty-slice signing
	// determinism: the root's nil slices are normalized when signed, then
	// Marshal/ParseBudgetToken round-trips them through JSON as non-nil
	// []string{}, and verification must still recompute an identical
	// signing payload from that reconstructed struct.
	token, delegatePriv, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: BudgetGrant{
		MaxTokens: 1_000_000,
		MaxDepth:  3,
		ExpiresAt: now.Add(time.Hour),
	}})
	if err != nil {
		t.Fatal(err)
	}

	delegated, _, err := Attest(token, delegatePriv, AttestOptions{Grant: BudgetGrant{
		MaxTokens: 500,
		MaxDepth:  0,
		ExpiresAt: now.Add(10 * time.Minute),
		// Providers/Models deliberately left nil here too.
	}})
	if err != nil {
		t.Fatal(err)
	}

	encoded, err := delegated.Marshal()
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}

	parsed, err := ParseBudgetToken(encoded)
	if err != nil {
		t.Fatalf("parse: %v", err)
	}

	grant, err := verifyChainAt(parsed, authorityPub, now)
	if err != nil {
		t.Fatalf("verify round-tripped token: %v", err)
	}
	if grant.MaxTokens != 500 {
		t.Fatalf("unexpected grant after round-trip: %+v", grant)
	}
}

func TestBudgetTokenMarshalRoundTripWithRestrictedProviders(t *testing.T) {
	authorityPub, authorityPriv := genKey(t)
	now := time.Date(2026, 7, 5, 12, 0, 0, 0, time.UTC)

	token, delegatePriv, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: rootGrant(now)})
	if err != nil {
		t.Fatal(err)
	}
	delegated, _, err := Attest(token, delegatePriv, AttestOptions{Grant: BudgetGrant{
		MaxTokens: 500,
		Providers: []string{"openai"}, // narrowed subset of the root's [openai, anthropic]
		Models:    []string{"gpt-4o"},
		MaxDepth:  0,
		ExpiresAt: now.Add(10 * time.Minute),
	}})
	if err != nil {
		t.Fatal(err)
	}

	encoded, err := delegated.Marshal()
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	parsed, err := ParseBudgetToken(encoded)
	if err != nil {
		t.Fatalf("parse: %v", err)
	}

	grant, err := verifyChainAt(parsed, authorityPub, now)
	if err != nil {
		t.Fatalf("verify round-tripped restricted token: %v", err)
	}
	if len(grant.Providers) != 1 || grant.Providers[0] != "openai" {
		t.Fatalf("providers lost or corrupted across round-trip: %+v", grant.Providers)
	}
}

func TestBudgetGrantValidateRejectsMissingExpiry(t *testing.T) {
	_, authorityPriv := genKey(t)
	_, _, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: BudgetGrant{MaxTokens: 100}})
	if err == nil {
		t.Fatal("expected validation error for a grant with no ExpiresAt")
	}
}

func TestBudgetGrantValidateRejectsNegativeDepth(t *testing.T) {
	_, authorityPriv := genKey(t)
	_, _, err := NewRootBudgetToken(authorityPriv, AttestOptions{Grant: BudgetGrant{
		MaxTokens: 100,
		MaxDepth:  -1,
		ExpiresAt: time.Now().Add(time.Hour),
	}})
	if err == nil {
		t.Fatal("expected validation error for negative MaxDepth")
	}
}

func TestIsSubsetOf(t *testing.T) {
	cases := []struct {
		name          string
		child, parent []string
		want          bool
	}{
		{"exact match", []string{"openai"}, []string{"openai"}, true},
		{"proper subset", []string{"openai"}, []string{"openai", "anthropic"}, true},
		{"not a subset", []string{"google"}, []string{"openai", "anthropic"}, false},
		{"empty child against restricted parent is a widening", nil, []string{"openai"}, false},
		{"empty parent handled by caller, not this helper", []string{"openai"}, nil, false},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if got := isSubsetOf(tc.child, tc.parent); got != tc.want {
				t.Fatalf("isSubsetOf(%v, %v) = %v, want %v", tc.child, tc.parent, got, tc.want)
			}
		})
	}
}

File: packages/sdk-node/test/budget-attestation.test.ts
import { describe, it, expect } from 'vitest';
import { generateKeyPairSync, type KeyObject } from 'node:crypto';
import {
  attest,
  isSubsetOf,
  newRootBudgetToken,
  parseBudgetToken,
  sign,
  verifyChain,
  verifyPresentation,
  type BudgetGrant,
} from '../src/index.js';

// Mirrors packages/sdk-go/budget_attestation_test.go scenario-for-scenario.

function genKey(): { publicKey: KeyObject; privateKey: KeyObject } {
  return generateKeyPairSync('ed25519');
}

// Fixed instant for every test that verifies through the injectable-now
// path (verifyChain(token, root, nowMs)). Tests that verify through a
// real-wall-clock path (verifyPresentation's default now) anchor to
// Date.now() instead — a hardcoded date there is a time bomb that starts
// failing once the calendar passes it (Go's test file had exactly that bug).
const FIXED_NOW = Date.UTC(2026, 6, 5, 12, 0, 0);

const HOUR_MS = 60 * 60 * 1000;
const MINUTE_MS = 60 * 1000;

function rootGrant(nowMs: number): BudgetGrant {
  return {
    maxTokens: 1_000_000,
    providers: ['openai', 'anthropic'],
    models: ['gpt-4o', 'claude-opus-4-5'],
    maxDepth: 3,
    expiresAt: new Date(nowMs + HOUR_MS),
  };
}

describe('budget attestation', () => {
  it('mints a root token and verifies the chain', () => {
    const authority = genKey();
    const { token, delegatePrivateKey } = newRootBudgetToken(authority.privateKey, { grant: rootGrant(FIXED_NOW) });

    expect(delegatePrivateKey).toBeDefined();
    expect(token.blocks).toHaveLength(1);

    const grant = verifyChain(token, authority.publicKey, FIXED_NOW);
    expect(grant.maxTokens).toBe(1_000_000);
  });

  it('rejects verification against the wrong root key', () => {
    const authority = genKey();
    const wrong = genKey();
    const { token } = newRootBudgetToken(authority.privateKey, { grant: rootGrant(FIXED_NOW) });

    expect(() => verifyChain(token, wrong.publicKey, FIXED_NOW)).toThrow(/invalid signature/);
  });

  it('narrows correctly across a single-hop delegation', () => {
    const authority = genKey();
    const root = newRootBudgetToken(authority.privateKey, { grant: rootGrant(FIXED_NOW) });

    const delegated = attest(root.token, root.delegatePrivateKey!, {
      grant: {
        maxTokens: 10_000, // narrower
        providers: ['openai'],
        models: ['gpt-4o'],
        maxDepth: 1, // narrower (<= 3-1)
        expiresAt: new Date(FIXED_NOW + 30 * MINUTE_MS),
      },
    });

    expect(delegated.delegatePrivateKey).toBeDefined();
    expect(delegated.token.blocks).toHaveLength(2);

    const grant = verifyChain(delegated.token, authority.publicKey, FIXED_NOW);
    expect(grant.maxTokens).toBe(10_000); // effective grant is the narrowed leaf
  });

  it('rejects a delegation that widens maxTokens', () => {
    const authority = genKey();
    const root = newRootBudgetToken(authority.privateKey, { grant: rootGrant(FIXED_NOW) });

    expect(() =>
      attest(root.token, root.delegatePrivateKey!, {
        grant: {
          maxTokens: 2_000_000, // wider than parent's 1,000,000
          maxDepth: 1,
          expiresAt: new Date(FIXED_NOW + 30 * MINUTE_MS),
        },
      }),
    ).toThrow(/does not narrow/);
  });

  it('rejects a delegation that widens the provider list', () => {
    const authority = genKey();
    const root = newRootBudgetToken(authority.privateKey, { grant: rootGrant(FIXED_NOW) });

    expect(() =>
      attest(root.token, root.delegatePrivateKey!, {
        grant: {
          maxTokens: 100,
          providers: ['openai', 'google'], // "google" not in parent's [openai, anthropic]
          maxDepth: 1,
          expiresAt: new Date(FIXED_NOW + 30 * MINUTE_MS),
        },
      }),
    ).toThrow(/does not narrow/);
  });

  it('rejects an empty provider list against a restricted parent (widening to "any")', () => {
    const authority = genKey();
    const root = newRootBudgetToken(authority.privateKey, { grant: rootGrant(FIXED_NOW) });

    expect(() =>
      attest(root.token, root.delegatePrivateKey!, {
        grant: {
          maxTokens: 100,
          providers: [], // "any provider" is looser than the parent's specific list
          models: ['gpt-4o'],
          maxDepth: 1,
          expiresAt: new Date(FIXED_NOW + 30 * MINUTE_MS),
        },
      }),
    ).toThrow(/does not narrow/);
  });

  it('rejects a delegation with a later expiry', () => {
    const authority = genKey();
    const root = newRootBudgetToken(authority.privateKey, { grant: rootGrant(FIXED_NOW) });

    expect(() =>
      attest(root.token, root.delegatePrivateKey!, {
        grant: {
          maxTokens: 100,
          maxDepth: 1,
          expiresAt: new Date(FIXED_NOW + 2 * HOUR_MS), // later than parent's 1 hour
        },
      }),
    ).toThrow(/does not narrow/);
  });

  it('rejects delegation from a depth-exhausted token', () => {
    const authority = genKey();
    const root = newRootBudgetToken(authority.privateKey, {
      grant: {
        maxTokens: 1000,
        maxDepth: 0, // no further delegation allowed
        expiresAt: new Date(FIXED_NOW + HOUR_MS),
      },
    });

    expect(() =>
      attest(root.token, root.delegatePrivateKey!, {
        grant: {
          maxTokens: 100,
          maxDepth: 0,
          expiresAt: new Date(FIXED_NOW + 30 * MINUTE_MS),
        },
      }),
    ).toThrow(/does not narrow/);
  });

  it('rejects attest with a private key that is not the current holder', () => {
    const authority = genKey();
    const wrong = genKey();
    const root = newRootBudgetToken(authority.privateKey, { grant: rootGrant(FIXED_NOW) });

    expect(() =>
      attest(root.token, wrong.privateKey, {
        grant: {
          maxTokens: 100,
          maxDepth: 1,
          expiresAt: new Date(FIXED_NOW + 30 * MINUTE_MS),
        },
      }),
    ).toThrow(/does not match the token's current holder key/);
  });

  it('rejects an expired token', () => {
    const authority = genKey();
    const { token } = newRootBudgetToken(authority.privateKey, { grant: rootGrant(FIXED_NOW) });

    const future = FIXED_NOW + 2 * HOUR_MS;
    expect(() => verifyChain(token, authority.publicKey, future)).toThrow(/expired/);
  });

  // Reproduces a real gap: the signature commits to expiresAt truncated to
  // whole seconds (see formatExpiryCanonical/signingPayload), but the
  // expiry CHECK used to compare against the raw, untruncated millisecond
  // value. A grant expiring at T+999ms was signed as if it expired at T,
  // but verifyChain would still accept it up to 999ms past T — the
  // enforced statement was looser than the signed one.
  it('enforces the same truncated expiry it signed, not the raw sub-second value', () => {
    const authority = genKey();
    const base = FIXED_NOW;
    const grant = { ...rootGrant(base), expiresAt: new Date(base + 999) };
    const { token } = newRootBudgetToken(authority.privateKey, { grant });

    // 500ms past `base`: before the RAW expiry (base+999ms), so the old
    // buggy check would have accepted this — but it's already after the
    // TRUNCATED expiry the signature actually committed to.
    expect(() => verifyChain(token, authority.publicKey, base + 500)).toThrow(/expired/);

    // At exactly `base` (the truncated/signed instant itself), it must
    // still verify — not yet expired.
    expect(() => verifyChain(token, authority.publicKey, base)).not.toThrow();
  });

  it('signs and verifies a presentation with the correct holder', () => {
    const authority = genKey();
    // verifyPresentation defaults to real wall-clock time, so the grant must
    // be anchored to Date.now(), not a fixed historical date.
    const now = Date.now();
    const { token, delegatePrivateKey } = newRootBudgetToken(authority.privateKey, { grant: rootGrant(now) });

    const context = Buffer.from('request-nonce-abc123');
    const signature = sign(token, delegatePrivateKey!, context);

    const grant = verifyPresentation(token, authority.publicKey, context, signature);
    expect(grant.maxTokens).toBe(1_000_000);
  });

  it('sign refuses a private key that does not match the holder key', () => {
    const authority = genKey();
    const impostor = genKey();
    const { token } = newRootBudgetToken(authority.privateKey, { grant: rootGrant(FIXED_NOW) });

    // The impostor doesn't hold the token's delegate key, so sign itself
    // must refuse — this is the defensive check inside sign.
    expect(() => sign(token, impostor.privateKey, Buffer.from('request-nonce'))).toThrow(
      /does not match the token's current holder key/,
    );
  });

  it('rejects a signature replayed under a different context', () => {
    const authority = genKey();
    // Same real-clock consideration as the presentation test above.
    const now = Date.now();
    const { token, delegatePrivateKey } = newRootBudgetToken(authority.privateKey, { grant: rootGrant(now) });

    const signature = sign(token, delegatePrivateKey!, Buffer.from('original-context'));
    expect(() =>
      verifyPresentation(token, authority.publicKey, Buffer.from('different-context'), signature),
    ).toThrow(/proof-of-possession signature invalid/);
  });

  it('issues to an explicit delegate public key without generating a keypair', () => {
    const authority = genKey();
    const explicitDelegate = genKey();

    const { token, delegatePrivateKey } = newRootBudgetToken(authority.privateKey, {
      grant: rootGrant(FIXED_NOW),
      delegatePublicKey: explicitDelegate.publicKey,
    });

    // No private key should be generated when delegatePublicKey was supplied.
    expect(delegatePrivateKey).toBeUndefined();

    // The explicit delegate's own private key must work.
    expect(() => sign(token, explicitDelegate.privateKey, Buffer.from('ctx'))).not.toThrow();
  });

  it('marshal round-trip preserves verification (nil-vs-empty lists)', () => {
    const authority = genKey();

    // Unrestricted root (omitted providers/models) so the child can validly
    // leave them out too — this exercises nil-vs-empty-list signing
    // determinism: the root's omitted lists are normalized when signed, then
    // marshal/parseBudgetToken round-trips them through JSON as [], and
    // verification must still recompute an identical signing payload from
    // that reconstructed grant.
    const root = newRootBudgetToken(authority.privateKey, {
      grant: {
        maxTokens: 1_000_000,
        maxDepth: 3,
        expiresAt: new Date(FIXED_NOW + HOUR_MS),
      },
    });

    const delegated = attest(root.token, root.delegatePrivateKey!, {
      grant: {
        maxTokens: 500,
        maxDepth: 0,
        expiresAt: new Date(FIXED_NOW + 10 * MINUTE_MS),
        // providers/models deliberately omitted here too.
      },
    });

    const encoded = delegated.token.marshal();
    const parsed = parseBudgetToken(encoded);

    const grant = verifyChain(parsed, authority.publicKey, FIXED_NOW);
    expect(grant.maxTokens).toBe(500);
  });

  it('marshal round-trip preserves a restricted provider list', () => {
    const authority = genKey();
    const root = newRootBudgetToken(authority.privateKey, { grant: rootGrant(FIXED_NOW) });

    const delegated = attest(root.token, root.delegatePrivateKey!, {
      grant: {
        maxTokens: 500,
        providers: ['openai'], // narrowed subset of the root's [openai, anthropic]
        models: ['gpt-4o'],
        maxDepth: 0,
        expiresAt: new Date(FIXED_NOW + 10 * MINUTE_MS),
      },
    });

    const parsed = parseBudgetToken(delegated.token.marshal());
    const grant = verifyChain(parsed, authority.publicKey, FIXED_NOW);
    expect(grant.providers).toEqual(['openai']);
  });

  it('a round-tripped token still supports proof-of-possession presentation', () => {
    const authority = genKey();
    const now = Date.now(); // presentation path checks real wall-clock time
    const { token, delegatePrivateKey } = newRootBudgetToken(authority.privateKey, { grant: rootGrant(now) });

    const parsed = parseBudgetToken(token.marshal());
    const context = Buffer.from('post-transport-nonce');
    const signature = sign(parsed, delegatePrivateKey!, context);

    const grant = verifyPresentation(parsed, authority.publicKey, context, signature);
    expect(grant.maxTokens).toBe(1_000_000);
  });

  it('validate rejects a grant with no expiry', () => {
    const authority = genKey();
    expect(() =>
      newRootBudgetToken(authority.privateKey, {
        grant: { maxTokens: 100, maxDepth: 0 } as unknown as BudgetGrant,
      }),
    ).toThrow(/expiresAt must be set/);
  });

  it('validate rejects negative maxDepth', () => {
    const authority = genKey();
    expect(() =>
      newRootBudgetToken(authority.privateKey, {
        grant: {
          maxTokens: 100,
          maxDepth: -1,
          expiresAt: new Date(Date.now() + HOUR_MS),
        },
      }),
    ).toThrow(/maxDepth must be >= 0/);
  });

  it('isSubsetOf table', () => {
    const cases: Array<{ name: string; child: string[] | null; parent: string[]; want: boolean }> = [
      { name: 'exact match', child: ['openai'], parent: ['openai'], want: true },
      { name: 'proper subset', child: ['openai'], parent: ['openai', 'anthropic'], want: true },
      { name: 'not a subset', child: ['google'], parent: ['openai', 'anthropic'], want: false },
      { name: 'empty child against restricted parent is a widening', child: null, parent: ['openai'], want: false },
      { name: 'empty parent handled by caller, not this helper', child: ['openai'], parent: [], want: false },
    ];
    for (const tc of cases) {
      expect(isSubsetOf(tc.child, tc.parent), tc.name).toBe(tc.want);
    }
  });
});
